### PR TITLE
[JENKINS-38034] SAML Plugin does not load groups when access with API Token

### PR DESCRIPTION
**Pull request by kuisathaverat**
fire loggedIn event when the user is logged to save the jenkins.security.LastGrantedAuthoritiesProperty this allow the plugin to restore authorities when you access by API token
also, I add a new user property to save creation date, last login, and a login counter

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ * Store details about create and login processes
+ *
+ * @author Kuisathaverat
+ */
+public class LoginDetailsProperty extends UserProperty {
+    private static final Logger LOG = Logger.getLogger(LoginDetailsProperty.class.getName());
+    private static final String ISO_8601 = "yyyy-MM-dd'T'HH:mm:ssZ";
+    private long createTimestamp;
+    private long lastLoginTimestamp;
+    private long loginCount;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: Supporting >2³¹ logins? :smiley:

File: src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+
+    /**
+     * Listen to the login success/failure event to persist {@link GrantedAuthority}s properly.
+     */
+    @Extension
+    public static class SecurityListenerImpl extends jenkins.security.SecurityListener {
+        @Override
+        protected void authenticated(@javax.annotation.Nonnull org.acegisecurity.userdetails.UserDetails details) {
+        }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: Huh, wonder why I made all these methods `abstract`. Should have left default no-op implementations.

File: src/main/resources/org/jenkinsci/plugins/saml/user/LoginDetailsProperty/config.jelly [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        software distributed under the License is distributed on an
+        "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+        KIND, either express or implied.  See the License for the
+        specific language governing permissions and limitations
+        under the License.
+-->
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+    <f:entry title="Login Details">
+        <p>Creation date: ${instance.getCreateDate()}</p>
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/saml/user/LoginDetailsProperty/config.jelly]: Note that `instance.createDate` suffices in JEXL.

File: src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java (hunk 1 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: yep, for tiresome users would be not enough for sure 🤣

**Comment by kuisathaverat** [comment]
>  Is there any meaningful event for failing to log in? If so, should fire that too.

I will review it in other PR

File: src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws FormException {
+        return this;
+    }
+
+
+    /**
+     * Listen to the login success/failure event to persist {@link GrantedAuthority}s properly.
+     */
+    @Extension
+    public static class SecurityListenerImpl extends jenkins.security.SecurityListener {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: surely a :bug:.  this should only be doing this IF the current security realm is SAML and not just if the SAML plugin is installed.
if you want to track this for every Authentication realm then it should be in a separate plugin

Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: @jtnord Is it enought?
` @Override
          protected void loggedIn(@javax.annotation.Nonnull String username) {
             SecurityRealm realm = Jenkins.getActiveInstance().getSecurityRealm();
             if(!(realm instanceof SamlSecurityRealm)) {
                 return;
             }`

Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/saml/user/LoginDetailsProperty.java]: Yup

**Pull request merged**